README.md
![Logo](admin/cec2.png)
### ioBroker.cec2

[![NPM version](http://img.shields.io/npm/v/iobroker.cec2.svg)](https://www.npmjs.com/package/iobroker.cec2)
[![Downloads](https://img.shields.io/npm/dm/iobroker.cec2.svg)](https://www.npmjs.com/package/iobroker.cec2)
![Number of Installations (latest)](http://iobroker.live/badges/cec2-installed.svg)
<!-- 
![Number of Installations (stable)](http://iobroker.live/badges/cec2-stable.svg)
-->
[![Dependency Status](https://status.david-dm.org/gh/iobroker-community-adapters/iobroker.cec2.svg)](https://david-dm.org/iobroker-community-adapters/iobroker.cec2)
[![Known Vulnerabilities](https://snyk.io/test/github/iobroker-community-adapters/ioBroker.cec2/badge.svg)](https://snyk.io/test/github/iobroker-community-adapters/ioBroker.cec2)

[![NPM](https://nodei.co/npm/iobroker.cec2.png?downloads=true)](https://nodei.co/npm/iobroker.cec2/)

**Tests:** ![Test and Release](https://github.com/iobroker-community-adapters/ioBroker.cec2/workflows/Test%20and%20Release/badge.svg)


Adapter for HDMI CEC

You can monitor / control devices using HDMI CEC. Most modern TVs and multimedia devices support CEC to some extent.

## Info

On start this adapter runs cec-client and polls all devices on the HDMI bus. For every CEC device a device in ioBroker
is created. The OSDName of a device will become its id in ioBroker (because it should not change and is nicely readable).
If devices show up during runtime they will be added to ioBroker, too.

#### CEC Adresses 

Short introduction about CEC addresses, you can skip this, but might want to read here, if you get confused by the description
below. 

In CEC there are two types of addresses. Both are important.

##### Logical Address

The logical address is a number between 0 and 15 (or F in Hex which is normally used). The number defines what kind of 
device it is: 
* 0 = TV
* 5 = Audiosystem
* 4,8,11 = Playback
* 1,2,9 = Recording
* 3,6,7,10 = Tuner

The others are reserved (12,13), Freeuse (14) and unregistered/broadcast (15). The devices that use these addresses are 
restricted in there communication.

The logical address is dynamically assigned. That means on one day you can have the situation that Playback A is assigned
address 4 and Playback B 8. But on the next day they are switched on in a different order and then A gets 8 and B 4. (Some devices
are always active on the CEC bus though and therefore cling to their addresses). If there are too many devices of one type
(i.e. 4 playback devices), then the logical addresses have to be reused and that will happen. The adapter tries to manage this case.

The logical address is used to address messages and reports to.
ioBroker needs a logical address, too, in order to receive reports from the bus. Therefore configure the adapter to a device type
that you know you have free logical addresses (for example recording). 

##### Physical Address

The physical address is based on the HDMI ports involved for the device. It is basically a path of port numbers towards 
the device. It is composed from 4 numbers seperated by dots. Some examples:
* 0.0.0.0 = TV
* 1.0.0.0 = 1. HDMI port of the TV
* 2.0.0.0 = 2. HDMI port of the TV
* 2.2.0.0 = 2. HDMI port of the TV, 2. HDMI port of a Switch / AV-Receiver
* 4.1.2.0 = 4. HDMI port of the TV, 1. HDMI port of a AV-Reciever / Switch, 2. HDMI port of the next AV-Receiver/Switch

That should give a basic idea. A 0 means the path is ending.

The physical address is fix for a device unless it is replugged into a different HDMI Port (or any devices before it seen from TV).

#### States per device 

For every device there are the following states created:

* info.active = device was seen recently and logical address should be correct
* info.cecVersion = should be 1.4, determines features
* info.lastSeen = last message from device on CEC bus
* info.logicalAddress = logical address as number
* info.logicalAddressHex = logical address as hex number (needed for own commands)
* info.Name = name the device sets in CEC Bus
* info.physicalAddress = physical address
* info.Vendor = vendor of the device
* activeSource = is the device the active source? Can switch this device to be active source
* menuStatus = lets device be controlled by TV remote
* state = power state of device (lets you switch it on/off, if supported)
* createButtons = press here to generate states for buttons in .buttons subfolder.
* buttons.time = time to press a button for (default 500ms).

## Buttons
Button presses do not work for all devices and some might need to have 
an active connection with the ioBroker device to be controlled via CEC bus.
For FireTV it works quite ok. 
To test button presses, press `createButtons` button in a device and test some of the
created buttons in some situations. Power works for quite a lot devices. 

#### Global States

There are the following global states:

* active-source = physical address of the current active source (can be set to switch inputs)
* arc = Audio Return Channel is (in)active might be able to (de)activate it here
* mute = mute Audiosystem
* raw-command = send commands to cec-client directly (for example 'scan' or tx + CEC Command from http://www.cec-o-matic.com/) 
* standbyAll = send all devices to standby
* systemAudio = Audiosystem is/is not in use. Informs devices that they should send volume/mute commands to Audiosystem
* volume = volume of Audiosystem, 0 = mute 
* volumeUp/Down = change volume of Audiosystem

#### Poll States

There is a subfolder "poll" with button-states for most states. If the button is triggered, the adapter will issue a 
command on CEC Bus to poll the value and set the state accordingly (sadly not all devices react to poll messages, though).

#### Requirements
cec-client have to be installed. Usually can be installed using:
```
sudo apt install cec-utils
```

The user running iobroker (nowadays "iobroker") needs acces to /dev/vchiq. You may need to add the iobroker user to the 
video group for that:
```
sudo usermod -a -G video iobroker
```

#### Installation
Execute the following command in the iobroker root directory (e.g. in /opt/iobroker)
```
npm install iobroker.cec2
```

Or install from admin webpage.


## Configuration

* osd name: this name will reported to other devices, for example your TV. You might need to select ioBroker there to receive remote controls in ioBroker.
* device type: You can set the device type, see discussion of logical address above to get an idea what that means. Use a device type that you have not many of.
* prevent unnamed devices: Sometimes devices are not reporting their name in certain situations (for example Nintendo Switch won't report its name when in standby but announce itself). In this situations the adapter might create a duplicate of the device under it's physical address (i.e. for numbers). You can enable this option to prevent it.  

## Script examples:

See [example Scripts](doc/ExampleScripts.md) for some example scripts that help with / repair multimedia setups.

## Changelog

<!--
	PLACEHOLDER for next version:
	## __WORK IN PROGRESS__
-->
### 0.1.0 (2021-02-18)
* add more translations.
* add more options (poll TV power).

### 0.0.8 (2021-02-14)
* Switched from unmaintained dependency to own code to control cec-monitor binary.
* Swtiched from event-stream to readline.
* Probably fixed missed incomming events.

### 0.0.7 (2021-01-24)
* fix warnings

### 0.0.6 (2021-01-02)
* update dependencies


### License
The MIT License (MIT)

Copyright (c) 2020-2021 Garfonso <garfonso@mobo.info>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
# Changelog

<!--
	PLACEHOLDER for next version:
	## __WORK IN PROGRESS__
-->
## __WORK IN PROGRESS__
* add more translations.

## 0.0.8 (2021-02-14)
* Switched from unmaintained dependency to own code to control cec-monitor binary.
* Swtiched from event-stream to readline.
* Probably fixed missed incomming events.

## 0.0.7 (2021-01-24)
* fix warnings

## 0.0.6 (2021-01-02)
* update dependencies

## 0.0.5 (2021-01-01)
* fix button presses
* add default for button press time

## 0.0.4 (2020-12-28)
* Make sure active devices are marked as active.
* make sure all devices have the required states
* fix deactivating devices
* Make sure we deliver incoming update if user did poll.

## 0.0.3 (2020-05-21)
* added 'preventUnnamedDevices' option ot prevent creation of devices that do not report their name. This sometimes happens if devices are talking on CEC bus but are not switched on (depends on device type).
* fixed possible crash on start

## 0.0.2 (2020-01-28)
* fixed a lot of bugs

## 0.0.1 (2020-01-28)
* initial release


lib/stateDefinitions.js
const CEC = require('./cec-constants');
const CECMonitor = require('./cec-monitor');

/**
 * stateDefinition, defines an ioBroker state and how it interacts with CEC
 * @typedef stateDefinition                 - defines an ioBroker state and how it interacts with CEC
 * @type {object}
 * @property {string} name                  - name of the state
 * @property {string} [key]                 - optional key of the state in stateDefinitions-Object,
 *                                               if name differs from key (because name is not suitable for ioBroker id)
 * @property {boolean} [readOnly]           - make readOnly for CEC
 * @property {function} [parse]             - parse CEC event to state
 * @property {string} [idPrefix]            - prefix id of state key with this string for subdirectories.
 * @property {number} [pollOpCode]          - opcode used for polling, will also create a poll-button.
 * @property {number} [pollArgument]        - argument for pollOpCode, if polling is more complex.
 * @property {number} [pollTarget]          - if polling targets a different logical address (especially for global states)
 * @property {function} [command]           - async function that is called if state is written -> i.e. control devices.
 * @property {boolean} [isGlobal]           - belongs to the global device
 *
 * @property {"string" | "number" | "boolean" | "object" | "array" | "mixed" | "file" | undefined} type - ioBroker type, i.e. "string", "number", "boolean", ...
 * @property {boolean} write                - make ioBroker state writable
 * @property {boolean} [read]               - make ioBroker state readable
 * @property {string} role                  - role in object.common
 * @property {string} [desc]                - description in object.common
 * @property {Record<string,number>} [valueList]    - valueList in object.common
 */

/**
 *
 * @type {Record<string, stateDefinition>}
 */
const stateDefinitions = {
    active: { //seen this session -> only then LE is valid.
        name: 'active',
        type: 'boolean',
        write: false,
        role: 'indicator.reachable',
        desc: 'device was active, logical address should be right',
        parse: () => true,
        idPrefix: 'info'
    },
    lastSeen: {
        name: 'lastSeen',
        'type': 'number',
        write: false,
        desc: 'last CEC command from device',
        role: 'value.time',
        parse: () => Date.now(),
        idPrefix: 'info'
    },
    //events exists for them:
    physicalAddress: {
        //bf:84:22:00:04
        //REPORT_PHYSICAL_ADDRESS: {"type":"TRAFFIC","number":"27086258","flow":"IN","source":11,"target":15,"opcode":132,"args":[34,0,4],"event":"REPORT_PHYSICAL_ADDRESS","data":{"val":8704,"str":"2.2.0.0"}}
        name: 'physicalAddress',
        type: 'string',
        desc: 'HDMI Ports on route from TV to device',
        role: 'info.address',
        write: false,
        pollOpCode: CEC.Opcode.GIVE_PHYSICAL_ADDRESS,
        idPrefix: 'info'
    },
    logicalAddress: {
        name: 'logicalAddress',
        type: 'number',
        role: 'info.address',
        write: false,
        desc: 'Logical Address (might change)',
        idPrefix: 'info'
    },
    logicalAddressHex: {
        name: 'logicalAddressHex',
        type: 'string',
        write: false,
        desc: 'Logical Address as Hex Number',
        role: 'info.address',
        idPrefix: 'info',
        readOnly: true
    },
    powerState: {
        //51:90:00
        //REPORT_POWER_STATUS: {"type":"TRAFFIC","number":"4584594","flow":"IN","source":5,"target":1,"opcode":144,"args":[0],"event":"REPORT_POWER_STATUS","data":{"val":0,"str":"ON"}}
        name: 'state',
        type: 'boolean',
        write: true,
        command: async function(value, device, cec) {
            if (value) {
                return cec.writeRawMessage('on ' + device.logicalAddress);
            } else {
                return cec.writeRawMessage('standby ' + device.logicalAddress);
            }
        },
        parse: (data) => {
            return data.data.val === CEC.PowerStatus.ON || data.data.val === CEC.PowerStatus.IN_TRANSITION_STANDBY_TO_ON;
        },
        role: 'switch.power',
        pollOpCode: CEC.Opcode.GIVE_DEVICE_POWER_STATUS,
        key: 'powerState'
    },
    activeSource: {
        //4f:82:21:00
        //ACTIVE_SOURCE: {"type":"TRAFFIC","number":"4602009","flow":"IN","source":4,"target":15,"opcode":130,"args":[33,0],"event":"ACTIVE_SOURCE","data":{"val":8448,"str":"2.1.0.0"}}
        name: 'activeSource',
        type: 'boolean',
        desc: 'switch input to this device and probably switches on TV',
        role: 'switch.enable',
        write: true,
        command: async function(value, device, cec) {
            if (value) {
                return cec.sendMessage(device.logicalAddress, CEC.LogicalAddress.BROADCAST, CEC.Opcode.ACTIVE_SOURCE, device.physicalAddress);
            } else {
                return cec.sendMessage(device.logicalAddress, CEC.LogicalAddress.TV, CEC.Opcode.INACTIVE_SOURCE, device.physicalAddress);
            }
        },
        parse: (data) => data.opcode === CEC.Opcode.ACTIVE_SOURCE,
        pollOpCode: CEC.Opcode.REQUEST_ACTIVE_SOURCE
    },
    recording: {
        name: 'recording',
        type: 'boolean',
        role: 'switch.record',
        write: true,
        command: async function(value, device, cec) {
            if (value) {
                return cec.sendMessage(null, device.logicalAddress, CEC.Opcode.RECORD_TV_SCREEN);
            } else {
                return cec.sendMessage(null, device.logicalAddress, CEC.Opcode.RECORD_OFF);
            }
        },

    },
    cecVersion: {
        //01:9e:05
        //CEC_VERSION: {"type":"TRAFFIC","number":"13890","flow":"IN","source":0,"target":1,"opcode":158,"args":[5],"event":"CEC_VERSION","data":{"val":5,"str":"VERSION_1_4"}}
        name: 'cecVersion',
        type: 'string',
        role: 'info.version',
        write: false,
        pollOpCode: CEC.Opcode.GET_CEC_VERSION,
        idPrefix: 'info'
    },
    /*language: {
        //0f:32:00:00:00
        //SET_MENU_LANGUAGE: {"type":"TRAFFIC","number":"4633220","flow":"IN","source":0,"target":15,"opcode":50,"args":[0,0,0],"event":"SET_MENU_LANGUAGE","data":{}}
        name: "language",
        type: "string",
        write: false,
        pollOpCode: CEC.Opcode.GET_MENU_LANGUAGE,
        idPrefix: "info"
    },*/
    menuStatus: {
        //51:8e:01
        //MENU_STATUS: {"type":"TRAFFIC","number":"23300","flow":"IN","source":5,"target":1,"opcode":142,"args":[1],"event":"MENU_STATUS","data":{}}
        //MENU_STATUS: {"type":"TRAFFIC","number":"5054","flow":"IN","source":11,"target":1,"opcode":142,"args":[0],"event":"MENU_STATUS","data":{}}
        name: 'menuStatus',
        type: 'boolean',
        write: true,
        role: 'switch.menu',
        command: async function(value, device, cec) {
            if (device.logicalAddress === CEC.LogicalAddress.TV) {
                await cec.sendMessage(null, device.logicalAddress, CEC.Opcode.MENU_REQUEST, value ? CEC.MenuRequestType.ACTIVATE : CEC.MenuRequestType.DEACTIVATE);
            }
            await cec.sendMessage(CEC.LogicalAddress.TV, device.logicalAddress, CEC.Opcode.MENU_REQUEST, value ? CEC.MenuRequestType.ACTIVATE : CEC.MenuRequestType.DEACTIVATE);
            return cec.sendMessage(null, device.logicalAddress, CEC.Opcode.MENU_REQUEST, CEC.MenuRequestType.QUERY); //query afterwards
        },
        parse: data => {
            return !data.args[0]; //0 = menu is active.
        },
        pollOpCode: CEC.Opcode.MENU_REQUEST,
        pollArgument: CEC.MenuRequestType.QUERY,
    },
    deck: {
        name: 'deck',
        type: 'number',
        write: true,
        role: 'media.state',
        valueList: CEC.DeckStatus,
        command: async function(value, device, cec, log) {
            let commandOpcode = CEC.Opcode.DECK_CONTROL;
            let argument;
            //might add "speed" state in order to let user choose min/max/medium speed options?
            switch (value) {
                case CEC.DeckStatus.STOP:
                    argument = CEC.DeckControl.STOP;
                    break;
                case CEC.DeckStatus.SKIP_FOWARD:
                case CEC.DeckStatus.INDEX_SEARCH_FOWARD:
                    argument = CEC.DeckControl.SKIP_FORWARD_WIND;
                    break;
                case CEC.DeckStatus.SKIP_REVERSE:
                case CEC.DeckStatus.INDEX_SEARCH_REVERSE:
                    argument = CEC.DeckControl.SKIP_REVERSE_REWIND;
                    break;
                case CEC.DeckStatus.NO_MEDIA:
                    argument = CEC.DeckControl.EJECT;
                    break;
                case CEC.DeckStatus.PLAY:
                    argument = CEC.Play.PLAY_FORWARD;
                    commandOpcode = CEC.Opcode.PLAY;
                    break;
                case CEC.DeckStatus.PLAY_REVERSE:
                    argument = CEC.Play.PLAY_REVERSE;
                    commandOpcode = CEC.Opcode.PLAY;
                    break;
                case CEC.DeckStatus.SLOW:
                    argument = CEC.Play.SLOW_FORWARD_MEDIUM_SPEED;
                    commandOpcode = CEC.Opcode.PLAY;
                    break;
                case CEC.DeckStatus.SLOW_REVERSE:
                    argument = CEC.Play.SLOW_REVERSE_MEDIUM_SPEED;
                    commandOpcode = CEC.Opcode.PLAY;
                    break;
                case CEC.DeckStatus.STILL:
                    argument = CEC.Play.PLAY_STILL;
                    commandOpcode = CEC.Opcode.PLAY;
                    break;
                case CEC.DeckStatus.FAST_FOWARD:
                    argument = CEC.Play.FAST_FORWARD_MEDIUM_SPEED;
                    commandOpcode = CEC.Opcode.PLAY;
                    break;
                case CEC.DeckStatus.FAST_REVERSE:
                    argument = CEC.Play.FAST_REVERSE_MEDIUM_SPEED;
                    commandOpcode = CEC.Opcode.PLAY;
                    break;
                case CEC.DeckStatus.RECORD:
                    commandOpcode = CEC.Opcode.RECORD_ON;
                    argument = null;
                    break;
                default:
                    log.info(value + ' not recognized as play command.');
                    return;
            }
            return cec.sendMessage(null, device.logicalAddress, commandOpcode, argument);
        },
        pollOpCode: CEC.Opcode.GIVE_DECK_STATUS,
        pollArgument: CEC.GiveStatus.ON //is on right? Or once? or off? Why do we need to set that parameter here?
    },
    tuner: {
        name: 'tuner',
        type: 'string',
        write: false,
        role: 'media.state',
        parse: function (data) {
            return JSON.stringify(data); //in order to gather more information...
        },
        pollOpCode: CEC.Opcode.GIVE_TUNER_DEVICE_STATUS,
        pollArgument: CEC.GiveStatus.ON //is on right? Or once? or off? Why do we need to set that parameter here?
    },
    vendor: {
        //4f:87:08:00:46
        //DEVICE_VENDOR_ID: {"type":"TRAFFIC","number":"4602390","flow":"IN","source":4,"target":15,"opcode":135,"args":[8,0,70],"event":"DEVICE_VENDOR_ID","data":{"val":524358,"str":"SONY"}}
        //DEVICE_VENDOR_ID: {"type":"TRAFFIC","number":"27086498","flow":"IN","source":11,"target":15,"opcode":135,"args":[0,0,0],"event":"DEVICE_VENDOR_ID","data":{"val":0,"str":"UNKNOWN"}}
        name: 'vendor',
        type: 'string',
        write: false,
        role: 'info.vendor',
        parse: function (data) {
            return data.data.str || 'Unknown (' + data.data.val + ')';
        },
        pollOpCode: CEC.Opcode.GIVE_DEVICE_VENDOR_ID,
        idPrefix: 'info'
    },
    name: {
        //51:47:52:58:2d:56:37:38:31
        //SET_OSD_NAME: {"type":"TRAFFIC","number":"22729","flow":"IN","source":5,"target":1,"opcode":71,"args":[82,88,45,86,55,56,49],"event":"SET_OSD_NAME","data":{"val":"RX-V781","str":"RX-V781"}}
        name: 'name',
        type: 'string',
        write: false,
        pollOpCode: CEC.Opcode.GIVE_OSD_NAME,
        idPrefix: 'info',
        role: 'info.name'
    },
    //uprocessed:
    //5f:81:22:00
    //ROUTING_INFORMATION: {"type":"TRAFFIC","number":"4649954","flow":"IN","source":5,"target":15,"opcode":129,"args":[34,0],"event":"ROUTING_INFORMATION","data":{}}
    //0f:80:00:00:20:00
    //ROUTING_CHANGE: {"type":"TRAFFIC","number":"4649953","flow":"IN","source":0,"target":15,"opcode":128,"args":[0,0,32,0],"event":"ROUTING_CHANGE","data":{"from":{"val":0,"str":"0.0.0.0"},"to":{"val":8192,"str":"2.0.0.0"}}}
    //41:00:70:00
    //undefined: {"type":"TRAFFIC","number":"4529265","flow":"IN","source":4,"target":1,"opcode":0,"args":[112,0]}


    //global states:
    volume: {
        //REPORT_AUDIO_STATUS: {"type":"TRAFFIC","number":"4495","flow":"IN","source":5,"target":1,"opcode":122,"args":[40],"event":"REPORT_AUDIO_STATUS","data":{}}
        name: 'volume',
        type: 'number',
        write: false,
        role: 'level.volume',
        parse: (data) => {
            const vol = parseInt(data.args.join(''), 16);
            const mute = 0x80 & vol;
            if (mute) {
                return 0;
            } else {
                return vol;
            }
        },
        command: async function(value, device, cec) {
            let button = CEC.UserControlCode.MUTE;
            if (value > 0) {
                button = value > device.volume ? CEC.UserControlCode.VOLUME_UP : CEC.UserControlCode.VOLUME_DOWN;
            }
            await cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_PRESSED, button);
            return cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_RELEASE);
        },
        pollOpCode: CEC.Opcode.GIVE_AUDIO_STATUS,
        pollTarget: CEC.LogicalAddress.AUDIOSYSTEM,
        isGlobal: true
    },
    systemAudio: {
        //SYSTEM_AUDIO_MODE_STATUS: {"type":"TRAFFIC","number":"4656","flow":"IN","source":5,"target":1,"opcode":126,"args":[1],"event":"SYSTEM_AUDIO_MODE_STATUS","data":{}}
        name: 'systemAudio',
        type: 'boolean',
        write: true,
        role: 'media.mode',
        command: async function(_value, _device, cec) {
            //maybe need to do SYSTEM_AUDIO_MODE_REQUEST from LE of active source and physical address of active source, too?
            return cec.sendCommand(cec.getActiveSource(), CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.SYSTEM_AUDIO_MODE_REQUEST, CECMonitor.EVENTS.SET_SYSTEM_AUDIO_MODE, cec.getActiveSourcePhysical());
        },
        parse: (data) => !!data.args[0],
        pollOpCode: CEC.Opcode.GIVE_SYSTEM_AUDIO_MODE_STATUS,
        pollTarget: CEC.LogicalAddress.AUDIOSYSTEM,
        isGlobal: true
    },
    arc: {
        name: 'arc',
        type: 'boolean',
        write: true,
        role: 'media.mode',
        parse: (data) => data.opcode === CEC.Opcode.REPORT_ARC_STARTED,
        command: async function(value, _device, cec) {
            //evaluate if that works
            return cec.sendMessage(CEC.LogicalAddress.AUDIOSYSTEM, CEC.LogicalAddress.TV, value ? CEC.Opcode.START_ARC : CEC.Opcode.END_ARC);
        },
        isGlobal: true
    },
    'raw-command': {
        name: 'raw-command',
        desc: 'send command to cec-client',
        type: 'string',
        write: true,
        role: 'text',
        command: async function (value, _device, cec) {
            await cec.writeRawMessage(value);
        },
        isGlobal: true
    },
    'active-source': {
        //ACTIVE_SOURCE: {"type":"TRAFFIC","number":"4861","flow":"IN","source":11,"target":15,"opcode":130,"args":[34,0],"event":"ACTIVE_SOURCE","data":{"val":8704,"str":"2.2.0.0"}}
        command: async function(value, device, cec) {
            if (!value) {
                return;
            }
            if (value.length === 1) { //allow single numbers for hdmi port.
                value += '.0.0.0';
            }
            let dev = device.devices.find(d => d.physicalAddress === value);
            if (!dev) {
                dev = device.devices.find(d => d.physicalAddress && d.physicalAddress.charAt(0) === value.charAt(0)); //if no full address, just search for first number.
            }
            let sender = null;
            if (dev) {
                sender = dev.logicalAddress;
            }
            await cec.sendMessage(sender, CEC.LogicalAddress.BROADCAST, CEC.Opcode.ACTIVE_SOURCE, value);
        },
        parse: (e) => e.data.str,
        isGlobal: true,
        name: 'active-source',
        desc: 'physical address of active playback device',
        type: 'string',
        role: 'media.input',
        write: true,
        read: true,
        pollOpCode: CEC.Opcode.REQUEST_ACTIVE_SOURCE
    },
    volumeUp: {
        name: 'volumeUp',
        desc: 'increase volume',
        type: 'boolean',
        write: true,
        read: false,
        role: 'button.volume.up',
        command: async function (_value, _device, cec) {
            await cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_PRESSED, CEC.UserControlCode.VOLUME_UP);
            await cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_RELEASE);
        },
        isGlobal: true
    },
    volumeDown: {
        name: 'volumeDown',
        desc: 'decrease volume',
        type: 'boolean',
        write: true,
        read: false,
        role: 'button.volume.down',
        command: async function (_value, _device, cec) {
            await cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_PRESSED, CEC.UserControlCode.VOLUME_DOWN);
            await cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_RELEASE);
        },
        isGlobal: true
    },
    mute: {
        name: 'mute',
        desc: 'mute audio',
        type: 'boolean',
        write: true,
        read: false,
        role: 'button.mute',
        command: async function (_value, _device, cec) {
            await cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_PRESSED, CEC.UserControlCode.MUTE);
            await cec.sendMessage(null, CEC.LogicalAddress.AUDIOSYSTEM, CEC.Opcode.USER_CONTROL_RELEASE);
        },
        isGlobal: true
    },
    standbyAll: {
        name: 'standbyAll',
        desc: 'switch off all CEC devices',
        type: 'boolean',
        write: true,
        read: false,
        role: 'button.power.off',
        command: async function (_value, _device, cec) {
            await cec.sendMessage(null, CEC.LogicalAddress.BROADCAST, CEC.Opcode.STANDBY);
        },
        isGlobal: true
    },
    //TODO: needs testing (!!) and needs to be implemented(??) -> especially: how to set time parameter?
    'osd-message': {
        name: 'osdMessage',
        desc: 'display osd message on screen',
        type: 'string',
        write: true,
        role: 'text',
        command: async function (value, _device, cec) {
            if (value) {
                const param = [CEC.DisplayControl.DISPLAY_FOR_DEFAULT_TIME].concat(value.split('').map(s => s.charCodeAt(0)));
                await cec.sendMessage(null, CEC.LogicalAddress.TV, CEC.Opcode.SET_OSD_STRING, param);
            }
        },
        isGlobal: true
    },
    'osd-message-clear': {
        name: 'osdClear',
        desc: 'clear previously send osd message',
        type: 'boolean',
        write: true,
        role: 'button',
        command: async function (_value, _device, cec) {
            await cec.sendMessage(null, CEC.LogicalAddress.TV, CEC.Opcode.SET_OSD_STRING, CEC.DisplayControl.CLEAR_PREVIOUS_MESSAGE);
        },
        isGlobal: true
    },
    'createButtons': {
        name: 'createButtons',
        desc: 'create states to send controls to this device. Not supported by all devices.',
        type: 'boolean',
        write: true,
        role: 'button'
    }
};

module.exports = stateDefinitions;
